File: flutter_app/lib/domain/services/pricing.dart
import '../models/enums.dart';
import '../models/owned_card.dart';

/// Cotation marché d'une carte, en centimes, par état et par gradation.
class MarketQuote {
  const MarketQuote({
    required this.cardId,
    required this.spotCents,
    required this.byCondition,
    required this.byGrade,
  });

  final String cardId;
  final int spotCents; // Near Mint brut de référence
  final Map<Condition, int> byCondition;
  final Map<String, int> byGrade; // ex: "PSA 10" -> cents
}

/// Service de cotation : règles pures de calcul de valeur. Aucune I/O.
abstract final class Pricing {
  /// Valeur courante (centimes) d'un exemplaire selon état/gradation.
  static int valueOf(OwnedCard owned, MarketQuote? quote) {
    if (quote == null) return 0;
    final base = quote.spotCents;
    if (owned.grade.isGraded) {
      final ladder = gradeMultipliers[owned.grade.company] ?? const {};
      final mult = ladder[owned.grade.score] ?? 1.0;
      return (base * mult).round();
    }
    return (base * owned.condition.multiplier).round();
  }

  /// Plus/moins-value latente d'un exemplaire.
  static ({int cents, double ratio}) unrealizedPnl(
    OwnedCard owned,
    MarketQuote? quote,
  ) {
    final current = valueOf(owned, quote);
    final cents = current - owned.acquiredCents;
    final ratio = owned.acquiredCents > 0 ? cents / owned.acquiredCents : 0.0;
    return (cents: cents, ratio: ratio);
  }
}

File: src/domain/services/pricing.ts
import { CONDITION_META, GRADE_MULTIPLIER } from '../taxonomy';
import { Grade, MarketQuote, OwnedCard } from '../types';

/**
 * Service de cotation : règles métier pures de calcul de valeur d'un
 * exemplaire à partir d'une cotation marché de référence. Aucune I/O.
 */

/**
 * Valeur courante (centimes) d'un exemplaire possédé, en tenant compte de
 * son état physique et de sa gradation.
 *
 * Logique : on part du spot Near Mint brut, on applique soit le
 * multiplicateur de gradation (si gradée), soit la décote d'état (si brute).
 */
export function valueOfOwnedCard(
  owned: OwnedCard,
  quote: MarketQuote | undefined,
): number {
  if (!quote) return 0;

  const base = quote.spotCents;

  if (owned.grade.company !== 'Raw' && owned.grade.score != null) {
    const ladder = GRADE_MULTIPLIER[owned.grade.company];
    const mult = ladder[owned.grade.score] ?? 1;
    return Math.round(base * mult);
  }

  const decote = CONDITION_META[owned.condition].multiplier;
  return Math.round(base * decote);
}

/** Libellé court d'une gradation pour les badges, ex: "PSA 10" ou "NM". */
export function gradeLabel(grade: Grade, condition: OwnedCard['condition']): string {
  if (grade.company === 'Raw' || grade.score == null) {
    return CONDITION_META[condition].short;
  }
  return `${grade.company} ${grade.score}`;
}

/**
 * Plus/moins-value latente d'un exemplaire : valeur courante - prix payé.
 */
export function unrealizedPnl(
  owned: OwnedCard,
  quote: MarketQuote | undefined,
): { cents: number; ratio: number } {
  const current = valueOfOwnedCard(owned, quote);
  const cents = current - owned.acquiredCents;
  const ratio = owned.acquiredCents > 0 ? cents / owned.acquiredCents : 0;
  return { cents, ratio };
}
